APE_public.rst

```diff
@@ -0,0 +1,284 @@
+applies to the module in which it is defined.  This means that a module can
+define an ``__all__`` attribute but if the module itself is not public, then
+anything in the ``__all__`` attribute cannot be publicly accessed from outside
+the module. For example, consider the following module::
+
+   src/
+      __init__/
+         __all__ = []
+
+      _foo/
+         __all__ = ['bar']
+         def bar():
+            pass
+
+In this case, ``bar`` is public in ``_foo``, but  ``src._foo`` is not public, so
+``src._foo.bar`` is not public either.
+
+We propose the following:
+
+1. That Astropy adopts all the PEP 8 rules on public versus internal interfaces
+   (the only one I'm not sure about is adding prefixes to private classes and
+   functions.)
+2. That Astropy disambiguates these rules by adopting the subsequent rules for
+   its API:
+3. That Astropy ensures its documentation is consistent with its code, the
+   latter being the authoritative source.
+
+   - That Astropy explains these rules in its documentation
+4. That Astropy also adopts these rules for coordinated packages, and recommends
+   that affiliated packages follow these rules as well.
+
+Rules:
+
+1. A symbol is public if all containing namespaces (packages, modules, classes,
+   functions, attributes or other names) are public. A symbol is private if any
+   containing namespace is private.
+2. All modules must have an ``__all__`` attribute, even if it is empty. The
+   ``__all__`` attribute defines the public and private interface of the module
+   in which it is defined. Anything in ``__all__`` is public, including
+   underscore-prefixed symbols. Anything not in ``__all__`` is private in that
+   module. The exception to this rule is implicit namespace packages, which
+   cannot have a top-level ``__all__`` attribute because they do not have a
+   top-level ``__init__.py`` file. In these cases, the public interface is
+   defined by the following rule.
+3. A symbol is public if it does not start with an underscore, and private
+   otherwise. The only exception to this rule is dunder symbols (``__<...>__``),
+   which follow their rules not determined in this APE.
+
+
+Branches and pull requests
+--------------------------
+
+N/A
+
+
+Implementation
+--------------
+
+This APE will be implemented in 3 phases:
+
+1. **Add** ``__all__``: An ``__all__`` dunder attribute will be added to all
+   modules that do not have it.
+2. **Update the documentation** to reflect the new rules.
+
+   - Add a scipy-like section to the developer documentation.
+   - Make sure the developer documentation is consistent with the new rules.
+   - Fix links to always point to the public interface, not the internal
+     interface.
+   - Clearly state if a documented object is actually private.
+3. **Add prefixes**: 1. Add prefixes to all modules that are not public. 2. Add
+   prefixes to all classes, functions, and attributes that are not public.
+      :note:`I'm less enthusiastic on this point.`
```

Comment:
I'm in favor of this as long as - as described above - we don't need to add underscore to symbols that are already in an underscore module for instance (so e.g. nothing in ``astropy.io.fits._tiled_compression`` needs to have an underscore prefix.

Reply:
I am fine with that. I have reservations about going all in on underscore prefixing everything. Not needing to underscore symbols in private modules (which still have an ``__all__``) SGTM.

Reply:
@astrofrog, resolve?